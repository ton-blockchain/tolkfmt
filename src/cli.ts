import {cac} from "cac"
import * as fs from "node:fs"
import * as path from "node:path"
import {glob} from "glob"
import {format, type Range} from "./index"

const version = "0.0.13"

type FormatMode = "format" | "format-and-write" | "check"

interface FormatFileOptions {
    readonly mode: FormatMode
    readonly range?: Range
    readonly sortImports: boolean
}

async function formatFile(
    filepath: string,
    options: FormatFileOptions,
): Promise<boolean | undefined> {
    const content = readFileOrFail(filepath)
    if (content === undefined) return undefined

    const {mode, range, sortImports} = options

    try {
        const [formattedCode, time] = await measureTime(async () => {
            return format(content, {maxWidth: 100, range, sortImports})
        })

        const alreadyFormatted = content === formattedCode

        if (mode === "check") {
            if (alreadyFormatted) {
                return true
            }
            console.log(`[warn]`, path.basename(filepath))
            return false
        }

        if (mode === "format-and-write") {
            console.log(
                path.basename(filepath),
                `${time.toFixed(0)}ms`,
                status(content, formattedCode),
            )
            fs.writeFileSync(filepath, formattedCode)
            return alreadyFormatted
        } else {
            process.stdout.write(formattedCode)
            return alreadyFormatted
        }
    } catch (error) {
        if (error instanceof Error) {
            console.error(
                `Cannot format file ${path.relative(process.cwd(), filepath)}:`,
                error.message,
            )
        }
        return undefined
    }
}

function status(before: string, after: string): string {
    if (before !== after) {
        return "(reformatted)"
    }
    return "(unchanged)"
}

async function measureTime<T>(fn: () => Promise<T>): Promise<[T, number]> {
    const startTime = performance.now()
    const result = await fn()
    const endTime = performance.now()
    const time = endTime - startTime
    return [result, time]
}

function readFileOrFail(filePath: string): string | undefined {
    try {
        return fs.readFileSync(filePath, "utf8")
    } catch (error: unknown) {
        if (error instanceof Error) {
            console.error(`Cannot read file: ${error.message}`)
        }
        return undefined
    }
}

function collectFilesToFormat(paths: string[]): string[] {
    return paths.flatMap(inputPath => {
        if (!fs.existsSync(inputPath)) {
            console.error(`Path does not exist: ${inputPath}`)
            return []
        }

        if (fs.statSync(inputPath).isFile()) {
            return inputPath
        }

        // for directory, find all .tolk files
        return glob.sync("**/*.tolk", {cwd: inputPath}).map(file => path.join(inputPath, file))
    })
}

interface CliOptions {
    readonly write: boolean | undefined
    readonly check: boolean | undefined
    readonly range: string | undefined
    readonly sortImports: boolean | undefined
}

function parseRange(rangeStr: string): Range {
    const regex = /^(\d+):(\d+)-(\d+):(\d+)$/
    const match = regex.exec(rangeStr)
    if (!match) {
        throw new Error(
            "Invalid range format. Expected: startLine:startChar-endLine:endChar (e.g., 1:5-3:10)",
        )
    }

    const [, startLine, startChar, endLine, endChar] = match
    return {
        start: {
            line: Number.parseInt(startLine, 10),
            character: Number.parseInt(startChar, 10),
        },
        end: {
            line: Number.parseInt(endLine, 10),
            character: Number.parseInt(endChar, 10),
        },
    }
}

export async function main(): Promise<void> {
    const cli = cac("tolkfmt")

    cli.version(version)
        .usage("[options] <files or directories>")
        .option("-w, --write", "Write result to same file")
        .option("-c, --check", "Check if the given files are formatted")
        .option(
            "-r, --range <range>",
            "Format only the specified range (format: startLine:startChar-endLine:endChar)",
        )
        .option("-s, --sort-imports", "Sort imports in the formatted file")
        .help()

    const parsed = cli.parse()
    // eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion
    const {write, check, range, sortImports = false} = parsed.options as CliOptions
    const filePaths = parsed.args

    if (write !== undefined && check !== undefined) {
        console.error("Error: Cannot use both --write and --check options together")
        process.exit(1)
    }

    let parsedRange: Range | undefined = undefined
    if (range !== undefined) {
        try {
            parsedRange = parseRange(range)
        } catch (error) {
            console.error("Error:", error instanceof Error ? error.message : "Invalid range format")
            process.exit(1)
        }
    }

    if (filePaths.length === 0) {
        return
    }

    const mode: FormatMode =
        check === undefined ? (write === undefined ? "format" : "format-and-write") : "check"

    if (mode === "check") {
        console.log("Checking formatting...")
    }

    const filesToFormat = collectFilesToFormat([...filePaths])

    if (filesToFormat.length === 0) {
        console.error("No .tolk files found")
        process.exit(1)
    }

    let someFileCannotBeFormatted = false
    let allFormatted = true

    for (const file of filesToFormat) {
        const res = await formatFile(file, {mode, range: parsedRange, sortImports})
        if (res === undefined) {
            someFileCannotBeFormatted = true
        } else {
            allFormatted &&= res
        }
    }

    if (check !== undefined) {
        if (allFormatted) {
            console.log("All Tolk files are properly formatted!")
        } else {
            console.log(
                "Code style issues found in the above files. Run tolkfmt with --write to fix.",
            )
            process.exit(1)
        }
    }

    if (someFileCannotBeFormatted) {
        process.exit(1)
    }
}

process.on("uncaughtException", error => {
    console.error("Unexpected error:", error.message)
    process.exit(1)
})

process.on("unhandledRejection", reason => {
    console.error("Unhandled promise rejection:", reason)
    process.exit(1)
})
