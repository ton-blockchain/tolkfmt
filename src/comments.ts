import type {Node as SyntaxNode, Node} from "web-tree-sitter"

export interface CommentInfo {
    readonly node: Node
    readonly start: number
    readonly end: number
    readonly startRow: number
    readonly endRow: number
    readonly text: string
}

// eslint-disable-next-line functional/type-declaration-immutability
export interface Bound {
    leading: CommentInfo[]
    trailing: CommentInfo[]
    dangling: CommentInfo[]
}

// eslint-disable-next-line functional/type-declaration-immutability
export type CommentMap = Map<number /* node.id */, Bound>

function isReallyTrailing(node: Node): boolean {
    const {nextSibling} = node
    if (nextSibling?.type === "comment") {
        // simple case
        return true
    }
    const nextNextSibling = nextSibling?.nextSibling
    if (nextSibling?.type === ",") {
        if (node.type === "call_argument") {
            // cannot step over comma in `foo(...)`
            return false
        }
    }
    if (nextNextSibling?.type === "comment") {
        return true
    }
    // likely not a trailing
    return false
}

export function bindComments(root: Node): CommentMap {
    const comments = collectComments(root)
    const byNode: CommentMap = new Map()

    const nodes = collectNamedNodes(root)

    // console.error("nodes:", nodes.map(it => "`" + it.text + `\` (${it.type})`).join("\n"))

    let index = 0

    outer: while (index < comments.length) {
        for (const node of nodes) {
            while (index < comments.length && comments[index].end <= node.startIndex) {
                attachLeading(comments[index++], node, byNode)
                continue outer
            }

            const lastTok = lastTokenRow(node)
            while (
                index < comments.length &&
                comments[index].start >= lastTok.endIndex &&
                comments[index].startRow === lastTok.endPosition.row &&
                isReallyTrailing(node)
            ) {
                attachTrailing(comments[index++], node, byNode)
                continue outer
            }

            while (
                index < comments.length &&
                comments[index].start > node.startIndex &&
                comments[index].end < node.endIndex &&
                !isInsideChild(comments[index].node, node)
            ) {
                attachDangling(comments[index++], node, byNode)
                continue outer
            }

            // Special handling for block_statement: comments inside empty blocks or between statements
            if (
                node.type === "block_statement" &&
                index < comments.length &&
                comments[index].start > node.startIndex &&
                comments[index].end < node.endIndex
            ) {
                const comment = comments[index]
                const statements = node.namedChildren.filter(
                    child => child !== null && child.type !== "comment",
                )

                // If block has no statements, comment should be dangling
                if (statements.length === 0) {
                    attachDangling(comments[index++], node, byNode)
                    continue outer
                }

                // Check if comment is between statements or not attached to any statement
                let attachedToStatement = false
                for (const stmt of statements) {
                    if (!stmt) continue
                    if (comment.end <= stmt.startIndex) {
                        // Comment is before this statement, should be leading to statement
                        attachedToStatement = true
                        break
                    }
                    if (comment.start >= stmt.endIndex) {
                        // Check if comment is on the same line as statement end - should be trailing
                        if (comment.startRow === stmt.endPosition.row) {
                            attachedToStatement = true
                            break
                        }
                        // Comment might be after this statement, continue checking
                        continue
                    }
                    if (comment.start >= stmt.startIndex && comment.end <= stmt.endIndex) {
                        // Comment is inside statement, let normal logic handle it
                        attachedToStatement = true
                        break
                    }
                }

                if (!attachedToStatement) {
                    attachDangling(comments[index++], node, byNode)
                    continue outer
                }
            }
        }

        const comment = comments[index]
        const parent = parentOfType(comment.node, "block_statement", "source_file")

        if (parent) {
            attachDangling(comment, parent, byNode)
        }

        // TODO: warn, we cannot attach comment!
        index++
    }

    while (index < comments.length) {
        attachTrailing(comments[index++], root, byNode)
    }

    return byNode
}

function lastTokenRow(node: Node): Node {
    let cur: Node | undefined = node
    while (cur !== undefined && cur.namedChildCount > 0) {
        cur = cur.namedChildren[cur.namedChildCount - 1] ?? undefined
    }
    return cur ?? node
}

export function getLeading(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id)
    if (!entry) return []
    return entry.leading
}

export function takeLeading(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id)
    if (!entry) return []
    const out = entry.leading
    entry.leading = []
    return out
}

export function takeTrailing(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id)
    if (!entry) return []
    const out = entry.trailing
    entry.trailing = []
    return out
}

export function takeDangling(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id)
    if (!entry) return []
    const out = entry.dangling
    entry.dangling = []
    return out
}

function dfs(n: Node, cb: (n: Node) => void): void {
    cb(n)
    for (const child of n.namedChildren) {
        if (!child) continue
        dfs(child, cb)
    }
}

function collectComments(root: Node): CommentInfo[] {
    const out: CommentInfo[] = []
    dfs(root, n => {
        if (n.type === "comment") {
            out.push({
                node: n,
                text: n.text,
                start: n.startIndex,
                end: n.endIndex,
                startRow: n.startPosition.row,
                endRow: n.endPosition.row,
            })
        }
    })
    return out.sort((a, b) => a.start - b.start)
}

function collectNamedNodes(root: Node): Node[] {
    const res: Node[] = []
    dfs(root, n => {
        if (n.isNamed && n.type !== "comment") res.push(n)
    })
    return res.sort((a, b) => a.startIndex - b.startIndex)
}

function ensureEntry(n: Node, map: CommentMap): Bound {
    let e = map.get(n.id)
    if (!e) {
        e = {leading: [], trailing: [], dangling: []}
        map.set(n.id, e)
    }
    return e
}

function attachLeading(c: CommentInfo, node: Node, map: CommentMap): void {
    // console.error(`attach leading ${c.text} to ${node.text}, id: ${node.id}, type: ${node.type}`)
    ensureEntry(node, map).leading.push(c)
}

function attachTrailing(c: CommentInfo, node: Node, map: CommentMap): void {
    // console.error(`attach trailing ${c.text} to ${node.text}, id: ${node.id}, type: ${node.type}`)
    ensureEntry(node, map).trailing.push(c)
}

function attachDangling(c: CommentInfo, node: Node, map: CommentMap): void {
    // console.error(`attach dangling ${c.text} to ${node.text}, id: ${node.id}, type: ${node.type}`)
    ensureEntry(node, map).dangling.push(c)
}

function isInsideChild(comment: Node, parent: Node): boolean {
    for (const child of parent.namedChildren) {
        if (!child) continue
        if (child.startIndex <= comment.startIndex && child.endIndex >= comment.endIndex) {
            return true
        }
    }
    return false
}

function parentOfType(node: SyntaxNode, ...types: readonly string[]): SyntaxNode | undefined {
    let {parent} = node

    for (let i = 0; i < 100; i++) {
        if (parent === null) return undefined
        if (types.includes(parent.type)) return parent
        const {parent: newParent} = parent
        parent = newParent
    }

    return undefined
}
