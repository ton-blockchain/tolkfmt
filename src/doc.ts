import type {Node} from "web-tree-sitter"
import type {CommentMap} from "./comments"

export type Doc =
    | Text
    | Line
    | SoftLine
    | HardLine
    | Indent
    | Group
    | Concat
    | LineSuffix
    | BreakParent
    | IfBreak
    | Empty

export interface Text {
    readonly $: "Text"
    readonly value: string
}

export const text = (value: string): Doc => ({
    $: "Text",
    value,
})

export interface Line {
    readonly $: "Line"
}

export const line = (): Doc => ({
    $: "Line",
})

export interface SoftLine {
    readonly $: "SoftLine"
}

export const softLine = (): Doc => ({
    $: "SoftLine",
})

export interface HardLine {
    readonly $: "HardLine"
}

export const hardLine = (): Doc => ({
    $: "HardLine",
})

export interface LineSuffix {
    readonly $: "LineSuffix"
    readonly suffix: Doc
}

export const lineSuffix = (suffix: Doc): Doc => ({
    $: "LineSuffix",
    suffix,
})

export interface BreakParent {
    readonly $: "BreakParent"
}

export const breakParent = (): Doc => ({
    $: "BreakParent",
})

export interface IfBreak {
    readonly $: "IfBreak"
    readonly breakContent: Doc | undefined
    readonly flatContent: Doc | undefined
}

export const ifBreak = (
    breakContent: Doc | undefined,
    flatContent: Doc | undefined = undefined,
): Doc => ({
    $: "IfBreak",
    breakContent,
    flatContent,
})

export interface Indent {
    readonly $: "Indent"
    readonly indent: number
    readonly content: Doc
}

export const indent = (content: Doc): Doc => ({
    $: "Indent",
    indent: 4,
    content,
})

export interface Group {
    readonly $: "Group"
    readonly content: Doc
}

export const group = (content: Doc[]): Doc => ({
    $: "Group",
    content: concat(content),
})

export interface Concat {
    readonly $: "Concat"
    readonly parts: Doc[]
}

export const concat = (parts: Doc[]): Doc => ({
    $: "Concat",
    parts,
})

export interface Empty {
    readonly $: "Empty"
}

export const empty = (): Doc => ({
    $: "Empty",
})

export function blankLinesBetween(a: Node, b: Node, comments: CommentMap): number {
    const aTrailing = comments.get(a.id)?.trailing ?? []
    const bLeading = comments.get(b.id)?.leading ?? []

    // If there are trailing comments after `a`, use the last trailing comment as the reference
    // If there are leading comments before `b`, use the first leading comment as the reference
    const {endPosition} = a
    const {startPosition} = b
    let endRow = endPosition.row
    let startRow = startPosition.row

    if (aTrailing.length > 0) {
        const lastTrailing = aTrailing.at(-1)
        if (lastTrailing) {
            const {endRow: lastEndRow} = lastTrailing
            endRow = lastEndRow
        }
    }

    if (bLeading.length > 0) {
        const [firstLeading] = bLeading
        const {startRow: firstStartRow} = firstLeading
        startRow = firstStartRow
    }

    const raw = startRow - endRow - 1
    return Math.max(0, raw)
}

export const blank = (n: number): Doc => (n === 0 ? hardLine() : concat([hardLine(), hardLine()]))
