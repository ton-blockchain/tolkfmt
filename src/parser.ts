import {Parser, Language} from "web-tree-sitter"

export let tolkLanguage: Language | undefined = undefined

export const initParser = async (treeSitterUri: string, tolkLangUri: string): Promise<void> => {
    if (tolkLanguage) {
        return
    }
    const options: object | undefined = {
        locateFile() {
            return treeSitterUri
        },
    }
    await Parser.init(options)
    tolkLanguage = await Language.load(tolkLangUri)
}

export function createTolkParser(): Parser {
    const parser = new Parser()
    if (tolkLanguage) {
        parser.setLanguage(tolkLanguage)
    }
    return parser
}
