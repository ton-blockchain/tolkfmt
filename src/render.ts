import type {Doc} from "./doc"
import {text} from "./doc"
import {empty} from "./doc"

export function render(doc: Doc, printWidth: number): string {
    type Mode = "flat" | "break"

    interface Frame {
        readonly doc: Doc
        readonly mode: Mode
        readonly indent: number
    }

    const out: string[] = []
    const lineSuffix: Doc[] = []
    const stack: Frame[] = [{doc, mode: "break", indent: 0}]

    function fits(d: Doc, w: number): boolean {
        const fitStack: Doc[] = [d]
        let width = w

        while (width >= 0 && fitStack.length > 0) {
            const cur = fitStack.pop() ?? empty()
            switch (cur.$) {
                case "Text": {
                    width -= cur.value.length
                    break
                }
                case "Line": {
                    width -= 1
                    break
                }
                case "SoftLine": {
                    break
                }
                case "HardLine": {
                    return false
                }
                case "Concat": {
                    for (let i = cur.parts.length - 1; i >= 0; i--) {
                        fitStack.push(cur.parts[i])
                    }
                    break
                }
                case "Indent": {
                    fitStack.push(cur.content)
                    break
                }
                case "Group": {
                    fitStack.push(cur.content)
                    break
                }
                case "LineSuffix": {
                    fitStack.push(cur.suffix)
                    break
                }
                case "BreakParent": {
                    return false
                }
                case "IfBreak": {
                    if (cur.flatContent) {
                        fitStack.push(cur.flatContent)
                    }
                    break
                }
                case "Empty": {
                    break
                }
            }
        }
        return width >= 0
    }

    // main loop
    while (stack.length > 0) {
        const {doc: cur, mode, indent} = stack.pop() ?? {doc: empty(), mode: "break", indent: 0}

        switch (cur.$) {
            case "Text": {
                if (cur.value === "\n") {
                    const prev = out.at(-1)
                    // eslint-disable-next-line @typescript-eslint/no-misused-spread
                    if (prev?.[0] === " " && ![...prev].some(it => it !== " ")) {
                        out[out.length - 1] = cur.value
                        break
                    }
                }
                out.push(cur.value)
                break
            }

            case "Line": {
                if (mode === "flat") {
                    out.push(" ")
                } else {
                    flushLineSuffix()
                    if (indent !== 0) {
                        stack.push({doc: text(" ".repeat(indent)), mode: "flat", indent: 0})
                    }
                    stack.push({doc: text("\n"), mode: "flat", indent: 0})
                }
                break
            }

            case "SoftLine": {
                if (mode !== "flat") {
                    flushLineSuffix()
                    if (indent !== 0) {
                        stack.push({doc: text(" ".repeat(indent)), mode: "flat", indent: 0})
                    }
                    stack.push({doc: text("\n"), mode: "flat", indent: 0})
                }
                break
            }

            case "HardLine": {
                flushLineSuffix()
                if (indent !== 0) {
                    stack.push({doc: text(" ".repeat(indent)), mode: "flat", indent: 0})
                }
                stack.push({doc: text("\n"), mode: "flat", indent: 0})
                break
            }

            case "Concat": {
                for (let i = cur.parts.length - 1; i >= 0; i--) {
                    stack.push({doc: cur.parts[i], mode, indent})
                }
                break
            }

            case "Indent": {
                stack.push({doc: cur.content, mode, indent: indent + cur.indent})
                break
            }

            case "Group": {
                const shouldFlat = fits(cur.content, printWidth - currentColumn(out))
                stack.push({doc: cur.content, mode: shouldFlat ? "flat" : "break", indent})
                break
            }

            case "LineSuffix": {
                lineSuffix.push(cur.suffix)
                break
            }

            case "BreakParent": {
                flushLineSuffix()
                break
            }

            case "IfBreak": {
                stack.push({
                    doc:
                        mode === "break"
                            ? (cur.breakContent ?? empty())
                            : (cur.flatContent ?? empty()),
                    mode,
                    indent,
                })
                break
            }

            case "Empty": {
                break
            }
        }
    }

    return out.join("")

    function flushLineSuffix(): void {
        while (lineSuffix.length > 0) {
            const suffix = lineSuffix.shift()
            if (suffix) {
                stack.push({doc: suffix, mode: "flat", indent: 0})
            }
        }
    }
}

function currentColumn(buf: string[]): number {
    let col = 0

    for (let i = buf.length - 1; i >= 0; i--) {
        const piece = buf[i]
        const nl = piece.lastIndexOf("\n")

        if (nl !== -1) {
            return col + piece.length - nl - 1
        }
        col += piece.length
    }
    return col
}
