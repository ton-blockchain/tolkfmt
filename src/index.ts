import {createTolkParser, initParser} from "./parser"
import {render} from "./render"
import {bindComments} from "./comments"
import type {Ctx} from "./print/ctx"
import {printNode} from "./print/node"
import type {Parser} from "web-tree-sitter"

export interface Position {
    readonly line: number
    readonly character: number
}

export interface Range {
    readonly start: Position
    readonly end: Position
}

export interface FormatOptions {
    readonly maxWidth?: number
    readonly parser?: Parser
    readonly range?: Range
    readonly sortImports?: boolean
}

export const format = async (code: string, opts?: FormatOptions): Promise<string> => {
    const {
        maxWidth = 100,
        parser = await createAndInitTolkParser(),
        range,
        sortImports = false,
    } = opts ?? {}

    const cst = parser.parse(code)

    const rootNode = cst?.rootNode
    if (!rootNode) return code

    const ctx: Ctx = {comments: bindComments(rootNode), range, sortImports}
    const doc = printNode(rootNode, ctx)
    if (!doc) return code

    return render(doc, maxWidth)
}

async function createAndInitTolkParser(): Promise<Parser> {
    await initParser(
        `${__dirname}/../wasm/tree-sitter.wasm`,
        `${__dirname}/../wasm/tree-sitter-tolk.wasm`,
    )
    return createTolkParser()
}
