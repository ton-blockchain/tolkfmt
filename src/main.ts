import {createTolkParser, initParser} from "./parser"
import {bindComments} from "./comments"
import {render} from "./render"
import type {Ctx} from "./print/ctx"
import {printNode} from "./print/node"

const main = async (): Promise<void> => {
    await initParser("../wasm/tree-sitter.wasm", "../wasm/tree-sitter-tolk.wasm")

    const parser = createTolkParser()

    const cst = parser.parse(`
fun main() {
     Foo<int32, int32>.new();
}
`)
    // bar(/* init: */ true, /* other: */ true /* other after */)

    if (!cst?.rootNode) throw new Error(`Unable to parse file`)

    const ctx: Ctx = {comments: bindComments(cst.rootNode), sortImports: true}
    const doc = printNode(cst.rootNode, ctx)

    if (doc) {
        console.log(render(doc, 50) + "---")
    }
}

void main()
