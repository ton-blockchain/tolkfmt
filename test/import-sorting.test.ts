import {format} from "../src"
import {initParser} from "../src/parser"

describe("Import sorting", () => {
    beforeAll(async () => {
        await initParser(
            `${__dirname}/../wasm/tree-sitter.wasm`,
            `${__dirname}/../wasm/tree-sitter-tolk.wasm`,
        )
    })

    it("should sort stdlib imports first", async () => {
        const code = `
import "./local"
import "@stdlib/hash"
import "other"
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should sort relative imports by category", async () => {
        const code = `
import "../parent"
import "./local"
import "nested/file"
import "../../grandparent"
import "./other_local"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should sort nested imports by depth", async () => {
        const code = `
import "nested/deep/very/deep"
import "nested/file"
import "nested/deep/file"
import "simple"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should sort parent imports by depth", async () => {
        const code = `
import "../../file"
import "../../../deep"
import "../parent"
import "../../../../very_deep"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should sort alphabetically within same category", async () => {
        const code = `
import "@stdlib/z"
import "@stdlib/a"
import "@stdlib/m"
import "./z"
import "./a"
import "./m"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should preserve comments with imports", async () => {
        const code = `
import "./local" // comment
import "@stdlib/hash"
// comment before
import "other"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should handle mixed import types", async () => {
        const code = `
import "nested/deep/file"
import "../parent"
import "@stdlib/string"
import "./local"
import "absolute"
import "@stdlib/array"
import "../../grandparent"
import "./other"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should support Windows paths", async () => {
        const code = `
import "nested\\deep\\file"
import ".\\local"
import "..\\parent"
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should treat paths without leading slash as relative", async () => {
        const code = `
import "simple_file"
import "nested/file"
import "./explicit_current"
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should keep inline comments", async () => {
        const code = `
import "simple_file" // comment 1
import "nested/file" // comment 2
import "./explicit_current" // comment 3
import "@stdlib/array" // comment 4

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should keep trailing comments", async () => {
        const code = `
// comment 1
import "simple_file"
// comment 2
import "nested/file"
// comment 3
import "./explicit_current"
// comment 4
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should treat paths with leading slash as absolute", async () => {
        const code = `
import "/absolute/path"
import "relative/path"
import "./current"
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: true})
        expect(result).toMatchSnapshot()
    })

    it("should not sort imports when flag is disabled", async () => {
        const code = `
import "./local"
import "@stdlib/hash"
import "other"
import "@stdlib/array"

fun test() {}`

        const result = await format(code, {sortImports: false})
        expect(result).toMatchSnapshot()
    })

    it("should not sort imports by default", async () => {
        const code = `
import "./local"
import "@stdlib/hash"
import "other"
import "@stdlib/array"

fun test() {}`

        const result = await format(code) // no sortImports flag
        expect(result).toMatchSnapshot()
    })
})
