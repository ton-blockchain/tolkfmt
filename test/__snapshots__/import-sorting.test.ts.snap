// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Import sorting should handle mixed import types 1`] = `
"import "@stdlib/array"
import "@stdlib/string"
import "./local"
import "./other"
import "absolute"
import "nested/deep/file"
import "../parent"
import "../../grandparent"

fun test() {}
"
`;

exports[`Import sorting should keep inline comments 1`] = `
"import "@stdlib/array" // comment 4
import "./explicit_current" // comment 3
import "simple_file" // comment 1
import "nested/file" // comment 2

fun test() {}
"
`;

exports[`Import sorting should keep trailing comments 1`] = `
"// comment 4
import "@stdlib/array"
// comment 3
import "./explicit_current"
// comment 1
import "simple_file"
// comment 2
import "nested/file"

fun test() {}
"
`;

exports[`Import sorting should not sort imports by default 1`] = `
"import "./local"
import "@stdlib/hash"
import "other"
import "@stdlib/array"

fun test() {}
"
`;

exports[`Import sorting should not sort imports when flag is disabled 1`] = `
"import "./local"
import "@stdlib/hash"
import "other"
import "@stdlib/array"

fun test() {}
"
`;

exports[`Import sorting should preserve comments with imports 1`] = `
"import "@stdlib/hash"
import "./local" // comment
// comment before
import "other"

fun test() {}
"
`;

exports[`Import sorting should sort alphabetically within same category 1`] = `
"import "@stdlib/a"
import "@stdlib/m"
import "@stdlib/z"
import "./a"
import "./m"
import "./z"

fun test() {}
"
`;

exports[`Import sorting should sort nested imports by depth 1`] = `
"import "simple"
import "nested/file"
import "nested/deep/file"
import "nested/deep/very/deep"

fun test() {}
"
`;

exports[`Import sorting should sort parent imports by depth 1`] = `
"import "../parent"
import "../../file"
import "../../../deep"
import "../../../../very_deep"

fun test() {}
"
`;

exports[`Import sorting should sort relative imports by category 1`] = `
"import "./local"
import "./other_local"
import "nested/file"
import "../parent"
import "../../grandparent"

fun test() {}
"
`;

exports[`Import sorting should sort stdlib imports first 1`] = `
"import "@stdlib/array"
import "@stdlib/hash"
import "./local"
import "other"

fun test() {}
"
`;

exports[`Import sorting should support Windows paths 1`] = `
"import "@stdlib/array"
import ".\\local"
import "nested\\deep\\file"
import "..\\parent"

fun test() {}
"
`;

exports[`Import sorting should treat paths with leading slash as absolute 1`] = `
"import "@stdlib/array"
import "./current"
import "relative/path"
import "/absolute/path"

fun test() {}
"
`;

exports[`Import sorting should treat paths without leading slash as relative 1`] = `
"import "@stdlib/array"
import "./explicit_current"
import "simple_file"
import "nested/file"

fun test() {}
"
`;
